### ISSUE TITLE
How to output long description

### ISSUE THREAD

**Issue by jkstyle2** [embodied-generalist/embodied-generalist#37]
Thanks for sharing your work.

Following the inference guide in README, I'm trying to inference for scene description task.
It seems working correct, but the output result is quite short.
I changed max_context_len, max_out_len in [vicuna7b.yaml](https://github.com/embodied-generalist/embodied-generalist/blob/main/configs/llm/vicuna7b.yaml), from 256 to 2048, but it still outputs very short as shown below.
Could you tell me how to extend the output length?

```
Loading checkpoint shards: 100%|
[2024-07-24 17:15:12,236][model.leo_agent][INFO] - Build Vicuna7B from ~/embodied-generalist/LEO_data/model_weights/weights_LLM/vicuna-7b
[2024-07-24 17:15:12,238][model.leo_agent][INFO] - Freeze LLM
[2024-07-24 17:15:12,894][timm.models._builder][INFO] - Loading pretrained weights from Hugging Face hub (('laion/CLIP-convnext_base_w-laion2B-s13B-b82K', 'open_clip_pytorch_model.bin'))
[2024-07-24 17:15:13,543][model.vision2d][INFO] - Build Backbone2D: convnext_base_laion2b, pretrain = True, freeze = True, pooling = avg
[2024-07-24 17:15:13,568][model.pcd_backbone][INFO] - Build PointcloudBackbone: PointNetPP
[2024-07-24 17:15:13,571][model.pcd_backbone][INFO] - Load PointNetPP weights from ~/embodied-generalist/LEO_data/model_weights/weights_3D_encoder/pointnetpp_vil3dref.pth
[2024-07-24 17:15:13,572][model.pcd_backbone][INFO] - Freeze PointNetPP
[2024-07-24 17:15:13,581][model.vision3d][INFO] - Build 3D module: OSE3D
[2024-07-24 17:15:13,598][model.leo_agent][INFO] - Apply LoRA with configs: {'flag': True, 'rank': 16, 'alpha': 16, 'target_modules': ['q_proj', 'k_proj', 'v_proj', 'o_proj', 'gate_proj', 'up_proj', 'down_proj'], 'dropout': 0.0}
[2024-07-24 17:15:59,711][inference][INFO] - Probe: load model from ~/embodied-generalist/LEO_data/model_weights/weights_leo
[2024-07-24 17:15:59,832][trainer.leo_trainer][INFO] - Successfully loaded from ~/embodied-generalist/LEO_data/model_weights/weights_leo, load_model_only = True
  0%|                                                                                                                                                                                                                                                                                                                                   | 0/1 [00:00<?, ?it/s]2024-07-24 17:16:03.869742: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 AVX512F AVX512_VNNI AVX512_BF16 AVX_VNNI AMX_TILE AMX_INT8 AMX_BF16 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-07-24 17:16:03.931943: I tensorflow/core/util/port.cc:104] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
[2024-07-24 17:16:04,754][inference][INFO] - {'source': 'scannet', 'scene_id': 'scene0601_00', 'situation': '', 'instruction': 'Describe this scene.', 'response': 'In this room, there is a red leather sofa. It is to the right of a brown wooden table. There are two picture hanging on the wall, one to the left and one to the right of each other. The ceiling is flat and white.'}
100%|
[2024-07-24 17:16:04,765] - Launched
```

**Comment by huangjy-pku** [comment]
The longest output during training is *scene captioning* and *task planning*. So the reason why the model cannot output a longer description is due to its inherent skill. And extending the `max_out_len` would not work because the output has not been truncated.

**Comment by jkstyle2** [comment]
The above result is from scene captioning. Do you mean the result is the longest output that LEO can make?
The annotation for scene captioning in train DB is not that short. 
Is it because LEO has trained with the limited output size on scene captioning task? 
How can I make the longer output as depicted in the paper below?
![image](https://github.com/user-attachments/assets/3d46ad74-90c2-408a-955c-171b96724bff)

**Comment by huangjy-pku** [comment]
If you want an expert in scene captioning (e.g., can produce longer responses), probably you should try the checkpoint without SFT (`align.pth`). The key is that SFT tasks are not so relevant to scene captioning and thus may degrade the capability of scene captioning.

**Comment by jkstyle2** [comment]
I tried scene captionin task with `align.pth`, but the result is worse than `sft_noact.pth`.
It even results out same outputs for different test scene as below.
I ran the same `scripts/inference.sh` with same yaml configs for the different tasks. Should I change some config or script in the code?
`[2024-07-26 16:31:51,344][inference][INFO] - {'source': 'scannet', 'scene_id': 'scene0603_00', 'situation': '', 'instruction': 'Describe this scene.', 'response': 'find the chair that is far away from the tv.'}`
`[2024-07-26 16:27:16,364][inference][INFO] - {'source': 'scannet', 'scene_id': 'scene0703_00', 'situation': '', 'instruction': 'Describe this scene.', 'response': 'find the chair that is farthest from the cabinet.'}`

**Comment by jkstyle2** [comment]
The reply for 'hello' is also different from what I expected. Is this the model's capability or am I doing incorrect?
`[2024-07-26 16:58:43,537][inference][INFO] - {'source': 'scannet', 'scene_id': 'scene0702_00', 'situation': '', 'instruction': 'hello', 'response': 'select the trash can that is farthest from the toilet paper.'}`

**Comment by jkstyle2** [comment]
@huangjy-pku 
Can I improve the accuracy of LEO for only scene description task by fine-tuning? For instance, from the pretrained checkpoint `align.pth`, I'd try fine-tuning on scene description task using only the dataset `3rscan_scenecap_train.json`.
or Would you suggest some other approaches for it?

**Comment by huangjy-pku** [comment]
I think other tasks may distract the learning of scene captioning skill. I believe finetuning solely on scene captioning task could mitigate this problem.

**Comment by jkstyle2** [comment]
Thanks for replying.
I'm currently trying to reproduce the entire process in the paper, by training the model from the beginning scratch.
I have only a single GPU(48G), so I modified `train_align_lora.sh` as 
```
python launch.py --mode python \
                 --name leo_align \
                 --qos lv0b \
                 --mem_per_gpu 100 \
                 --time 48 \
                 --config configs/default.yaml \
                 --port 2020 \
                 --gpu_per_node 4 \
                 --num_nodes 1 \
                 --partition HGX \
                 task=align \
                 note=align_lora \
```

also in `configs/default.yaml` , I changed batchsize to 3 (default) because of the out of momory size. 
In your paper, you're using 4 GPUs with 4 batchsize for 80GB memory size.
Can you tell me how long it takes to train with the configurations? Also, why did you choose batchsize=4 for your config? I think you could have set more batch size.

**Comment by huangjy-pku** [comment]
For the first stage, it takes 2~3 days with 4 × A100. The second stage uses more GPU memory due to the long context length and response length of dialogue and planning. So, we select batchsize=4 from the exponential list [1,2,4,8,...], and keep this consistent across the two stages.
